### PR TITLE
[Impeller] Reland 3: Implement draw order optimization.

### PR DESCRIPTION
**Pull request by bdero**
Attempt 1: #54136
Revert 1: #54067
Attempt 2: #54215
Revert 2: #54261
Attempt 3: #54268
Revert 3: #54325

This time for sure!

Includes a fix for the bug discovered in yt.

This turned out to be a bug that was laying dormant in EntityPass for about 5 months. During the clip replay, the clip scissor was getting set to `clip_coverage_stack.CurrentClipCoverage()` instead of the coverage stored along side the clip being replayed.

## Description
For each clip scope, draw opaque items in reverse order and translucent/backdrop-independent items in their original order afterwards. Clips are treated as translucent by the parent scope.

Respects clips, subpass collapse, and the clear color optimization.

### Local new_gallery before/after (iPhone 12 mini):
```
cd ~/projects/flutter/flutter/dev/integration_tests/new_gallery
flutter drive --profile --local-engine-src-path ~/projects/flutter/engine/src --local-engine=ios_profile --local-engine-host=host_profile_arm64 --trace-startup -t test_driver/transitions_perf.dart -d 00008101-000A59A93C10001E
```
image

### CODE REVIEW

File: impeller/entity/draw_order_resolver.cc [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+void DrawOrderResolver::PushClip(size_t element_index) {
+  draw_order_layers_.back().dependent_elements.push_back(element_index);
+  draw_order_layers_.push_back({});
+};
+
+void DrawOrderResolver::PopClip() {
+  if (draw_order_layers_.size() == 1u) {
+    // This is likely recoverable, so don't assert.
+    VALIDATION_LOG
+        << "Attemped to pop the first draw order clip layer. This is a bug in "
```
Comment (flar) [inline impeller/entity/draw_order_resolver.cc:31]: Should that read "may be a bug"?

**Review by jonahwilliams — COMMENTED** [review]
As long as the draw order resolver is working off of indices, it should be straightforward to port to the new canvas. That said, I plan to reland the switch after  https://github.com/flutter/engine/pull/54604 which would disable this optimization, unless you update exp canvas as well.

**Comment by flar** [comment]
> As long as the draw order resolver is working off of indices, it should be straightforward to port to the new canvas. That said, I plan to reland the switch after #54604 which would disable this optimization, unless you update exp canvas as well.

I'm still trying to reland indices (https://github.com/flutter/engine/pull/54676) so we can't do reorder optimization until that happens. And expCanvas doesn't use indices at all yet, so we'd need to decide if we want to do that conversion as a separate step or combine it with the re-order re-write.

File: impeller/entity/draw_order_resolver.cc (hunk 1 above)
Comment (bdero) [reply impeller/entity/draw_order_resolver.cc]: Done.

**Comment by bdero** [comment]
It's hard to know exactly what features we need in the DL until this actually sticks. The nature of what we'd need to do in exp canvas to support this kind of thing changed significantly from the first time I tried to land this.

So I'll help unravel the puzzle _after_ this change sticks.

File: impeller/entity/entity_pass.cc [hunk 2]
```diff
@@ -776,15 +846,19 @@ bool EntityPass::RenderElement(Entity& element_entity,
...
-      }
+    clip_coverage_stack.ActivateClipReplay();
+  }
+
+  // If there are any pending clips to replay, render any that may affect
+  // the entity we're about to render.
+  while (const EntityPassClipStack::ReplayResult* next_replay_clip =
+             clip_coverage_stack.GetNextReplayResult(element_entity)) {
+    auto& replay_entity = next_replay_clip->entity;
+    SetClipScissor(next_replay_clip->clip_coverage, *result.pass,
```
Comment (bdero) [inline impeller/entity/entity_pass.cc:857]: This is the only line that changed from the previous attempt (other than the new golden that I added).

`clip_coverage_stack.CurrentClipCoverage()` changed to `next_replay_clip->clip_coverage`.

**Review by chinmaygarde — APPROVED** [review]

**Pull request merged**